### PR TITLE
(DIO-2675) Undo pool size & template overrides

### PR DESCRIPTION
**Pull request by genebean**
This implements a delete method for pooltemplate and poolsize. The API removes the override from Redis and then adds an entry in Redis that causes the pool manager to wake up and process the removal of the override.

To facilitate this, a new variable has been created in lib/vmpooler.rb to hold a copy of the original / pre-override config. This supplemental copy of the pools is then indexed for use as a reference.

When pool manager wakes up to process an override removal, it looks up the pre-override value from the config via the new variables mentioned above.

Just as with entering overrides, no restart is needed. Template and pool size changes are logged so that anyone watching or reviewing the logs can see what happened when. The new API endpoints also return values for both the pre-revert and post-revert value.

There is a second commit that also makes it so that Dockerfile_local can use cached layers when all that is changing is VMPooler's code, and not its gems.

### CODE REVIEW

File: lib/vmpooler.rb [hunk 1]
```diff
@@ -133,8 +133,18 @@ def self.config(filepath = 'vmpooler.yaml')
       parsed_config[:pools] = load_pools_from_redis(redis)
     end
 
+    # Marshal.dump is paired with Marshal.load to create a copy that has its own memory space
+    # so that each can be edited independently
+    # rubocop:disable Security/MarshalLoad
+
+    # retain a copy of the pools that were observed at startup
+    serialized_pools = Marshal.dump(parsed_config[:pools])
+    parsed_config[:pools_at_startup] = Marshal.load(serialized_pools)
```
[the remark this answers is not in the document]
Comment (genebean) [reply lib/vmpooler.rb:142]: Clone doesn't work. 

```ruby
2.7.0 :025 >puts JSON.pretty_generate(parsed_config[:pools][0])
{
  "name": "almalinux-8-x86_64-pixa4",
  "alias": [
    "almalinux-8-x86_64"
  ],
  "template": "foo",
  "folder": "gene/acceptance4/almalinux-8-x86_64",
  "datastore": "vmpooler_acceptance4",
  "clone_target": "acceptance4",
  "size": 2,
  "datacenter": "pix",
  "provider": "vsphere",
  "snapshot_mainMem_ioBlockPages": "2048",
  "snapshot_mainMem_iowait": "5",
  "network": "vmpooler2"
}
 => nil 
2.7.0 :026 > parsed_config[:pools][0]['template']
 => "templates/netapp/acceptance4/almalinux-8-x86_64-0.0.2" 
2.7.0 :027 > parsed_config[:pools_at_startup][0]['template']
 => "templates/netapp/acceptance4/almalinux-8-x86_64-0.0.2" 
2.7.0 :028 > parsed_config[:pools][0]['template'] = 'foo'
 => "foo" 
2.7.0 :029 > parsed_config[:pools][0]['template']
 => "foo" 
2.7.0 :030 > parsed_config[:pools_at_startup][0]['template']
 => "foo"
```

File: spec/integration/api/v1/config_spec.rb [hunk 2]
```diff
@@ -26,6 +26,10 @@ def app()
         {'name' => 'pool1', 'size' => 5, 'template' => 'templates/pool1', 'clone_target' => 'default_cluster'},
         {'name' => 'pool2', 'size' => 10}
       ],
+      pools_at_startup: [
```
[the remark this answers is not in the document]
Comment (genebean) [reply spec/integration/api/v1/config_spec.rb:29]: I was trying to implement this and am now realizing I am not sure what you are suggesting.

File: docs/API.md [hunk 3]
```diff
@@ -743,6 +743,29 @@ $ curl -X POST -H "Content-Type: application/json" -d '{"debian-7-i386":"2","deb
...
+
+```
+$ curl -X DELETE -u jdoe --url vmpooler.example.com/api/v1/poolsize/almalinux-8-x86_64
+Enter host password for user 'jdoe':
+```
+```json
+{
+    "ok": true,
+    "pool_size_before_overrides": 2,
+    "pool_size_before_reset": 4
```
[the remark this answers is not in the document]
Comment (genebean) [reply docs/API.md:764]: I think I'd like to stay with what is here as I think it is more clear and is accurate regardless of if the pool size goes up or down as a result of calling this method.

File: docs/API.md [hunk 4]
```diff
@@ -775,6 +798,29 @@ $ curl -X POST -H "Content-Type: application/json" -d '{"debian-7-i386":"templat
...
+
+```
+$ curl -X DELETE -u jdoe --url vmpooler.example.com/api/v1/pooltemplate/almalinux-8-x86_64
+Enter host password for user 'jdoe':
+```
+```json
+{
+    "ok": true,
+    "template_before_overrides": "templates/almalinux-8-x86_64-0.0.2",
+    "template_before_reset": "templates/almalinux-8-x86_64-0.0.3-beta"
```
[the remark this answers is not in the document]
Comment (genebean) [reply docs/API.md:818]: I think I'd like to stay with what is here as I think it is more clear and is accurate regardless of if the pool size goes up or down as a result of calling this method.

File: lib/vmpooler/api/v1.rb [hunk 5]
```diff
@@ -289,6 +293,33 @@ def update_user_metrics(operation, vmname)
       puts 'd', "[!] [#{poolname}] failed while evaluating usage labels on '#{vmname}' with an error: #{e}"
     end
 
+    def reset_pool_size(poolname)
+      result = { 'ok' => false }
+
+      pool_index_live = pool_index(pools)
```
[the remark this answers is not in the document]
Comment (genebean) [reply lib/vmpooler/api/v1.rb:299]: Is there anything else you'd like me to do with this bit?

**Comment by genebean** [comment]
@sbeaulie I think this is ready for another review.

File: lib/vmpooler/api/v1.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbeaulie) [reply lib/vmpooler/api/v1.rb:299]: if the contents are the same we could use just one variable.

File: spec/integration/api/v1/config_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbeaulie) [reply spec/integration/api/v1/config_spec.rb:29]: I was thinking something adding to the existing context 'when no config is given' because it calls Vmpooler.config which is the code block where the Marshall load is happening. Then its easy to swap for .clone and see if it works to replace Marshall load (it does not as you also found out, because its an array and you need a deep clone for the array)
Meaning it actually tests the code block here https://github.com/puppetlabs/vmpooler/blob/master/lib/vmpooler.rb#L30

So the spec would add something like this:
```
    context 'when no config is given' do
      it 'keeps a copy of the original pools at startup' do
        Dir.chdir(fixtures_dir) do
          configuration = Vmpooler.config
          expect(configuration[:pools]).to eq(configuration[:pools_at_startup])
        end
      end
      it 'the copy is a separate object and not a reference' do
        Dir.chdir(fixtures_dir) do
          configuration = Vmpooler.config
          configuration[:pools][0]['template'] = 'sam'
          expect(configuration[:pools]).not_to eq(configuration[:pools_at_startup])
        end
      end
    end
```

File: lib/vmpooler/api/v1.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (genebean) [reply lib/vmpooler/api/v1.rb:299]: I just looked at the api docs and it appears that entire pools cannot be added or removed via it so I think it is safe to make this a single variable. I had done two different ones because I thought you might be able to change the actual list of pools post-startup.

[the remark this answers is not in the document]
Comment (genebean) [reply lib/vmpooler/api/v1.rb]: Fixed. I am also going to do a different pr to remove all the calls to the pool_index method as the index is created in lib/vmpooler.rb and shouldn't need to be recreated constantly

File: spec/integration/api/v1/config_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (genebean) [reply spec/integration/api/v1/config_spec.rb:29]: Thanks! Added.

File: spec/unit/vmpooler_spec.rb [hunk 6]
```diff
@@ -20,6 +20,21 @@
...
+        Dir.chdir(fixtures_dir) do
+          configuration = Vmpooler.config
+          expect(configuration[:pools]).to eq(configuration[:pools_at_startup])
+        end
+      end
+
+      it 'the copy is a separate object and not a reference' do
+        Dir.chdir(fixtures_dir) do
+          configuration = Vmpooler.config
+          configuration[:pools][0]['template'] = 'sam'
```
Comment (sbeaulie) [inline spec/unit/vmpooler_spec.rb:34]: lolz

File: lib/vmpooler.rb [hunk 7]
```diff
@@ -133,8 +133,17 @@ def self.config(filepath = 'vmpooler.yaml')
       parsed_config[:pools] = load_pools_from_redis(redis)
     end
 
+    # Marshal.dump is paired with Marshal.load to create a copy that has its own memory space
+    # so that each can be edited independently
+    # rubocop:disable Security/MarshalLoad
+
+    # retain a copy of the pools that were observed at startup
+    serialized_pools = Marshal.dump(parsed_config[:pools])
```
Comment (sbeaulie) [inline lib/vmpooler.rb:141]: could do one liner here
```suggestion
    parsed_config[:pools_at_startup] = Marshal.load(Marshal.dump(parsed_config[:pools]))
```

Comment (genebean) [reply lib/vmpooler.rb:141]: It think the one liner is a little hard to follow, personally

File: docker/Dockerfile_local [hunk 8]
```diff
@@ -21,9 +18,18 @@ RUN apt-get update -qq && \
     apt-get autoremove -y && \
     rm -rf /var/lib/apt/lists/*
 
+COPY docker/docker-entrypoint.sh /usr/local/bin/
+COPY ./Gemfile* ./
```
Comment (mattkirby) [inline docker/Dockerfile_local:22]: What is your hope of specifying the Gemfile like this? We explicitly exclude the lock file in .dockerignore.

File: lib/vmpooler/api/v1.rb [hunk 9]
```diff
@@ -28,6 +28,10 @@ def pools
       Vmpooler::API.settings.config[:pools]
     end
 
+    def pools_at_startup
```
Comment (mattkirby) [inline lib/vmpooler/api/v1.rb:31]: Nitpick, but is this meant to be pools loaded from config file, or all pools available at startup?

File: lib/vmpooler/api/v1.rb [hunk 10]
```diff
@@ -289,6 +293,32 @@ def update_user_metrics(operation, vmname)
...
+
+      pools_updated = 0
+      sync_pool_sizes
+
+      pool_size_now = pools[pool_index[poolname]]['size']
+      pool_size_original = pools_at_startup[pool_index[poolname]]['size']
+      result['pool_size_before_reset'] = pool_size_now
+      result['pool_size_before_overrides'] = pool_size_original
+
+      unless pool_size_now.to_i == pool_size_original.to_i
```
Comment (mattkirby) [inline lib/vmpooler/api/v1.rb:309]: One suggestion is to consider specifying 'to_i' when assigning pool_size_now and pool_size_original. Then, comparison does not have to specify to_i and simplifies reading that block.

**Review by mattkirby — APPROVED** [review]

File: docker/Dockerfile_local (hunk 8 above)
Comment (genebean) [reply docker/Dockerfile_local:22]: Missed that, I used a line I had from another file.

File: lib/vmpooler/api/v1.rb (hunk 10 above)
Comment (genebean) [reply lib/vmpooler/api/v1.rb:309]: Good call
